amaranth/vendor/_gatemate.py

```diff
@@ -0,0 +1,120 @@
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.ccf": r"""
+            # {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                Net "{{port_name}}" Loc = "{{pin_name}}"
+                {%- set constraints = [
+                    "SCHMITT_TRIGGER", "PULLUP", "PULLDOWN", "KEEPER", "SLEW", "DRIVE", 
+                    "DELAY_OBF", "DELAY_IBF", "FF_IBF", "FF_OBF", "LVDS_BOOST", "LVDS_TERM"
+                ] -%}
+                {%- for constraint in constraints if constraint in attrs -%}
+                    | {{constraint}}={{attrs[constraint]}}
+                {%- endfor -%};
+            {% endfor %}
+        """,
+        "{{name}}.sdc": r"""
+            # {{autogenerated}}
+            {% for net_signal, port_signal, frequency in platform.iter_clock_constraints() -%}
+                {% if port_signal is not none -%}
+                    create_clock -period {{1000000000/frequency}} [get_ports {{port_signal.name|tcl_quote}}]
+                {% endif %}
+            {% endfor %}
+        """,
+    }
+
+    command_templates = [
+        r"""
+        mkdir -p log;
+        mkdir -p net;
+        {{invoke_tool("yosys")}}
+            -ql log/synth.log -p 'read -sv {{name}}.v; synth_gatemate -top {{name}} -nomx8 -vlog net/{{name}}_synth.v'
+        """,
+        r"""
+
+        {{invoke_tool("p_r")}}
+            -v
+            -i net/{{name}}_synth.v
+            -o {{name}}
+            -ccf {{name}}.ccf
+            -cCP > log/impl.log
+        """,
+    ]
+
+    # Common logic
+
+    @property
+    def default_clk_constraint(self):
+        if self.default_clk == "sys_clk0":
+            return Clock(self.osc_freq / self.osc_div)
+        return super().default_clk_constraint
+
+    def add_clock_constraint(self, clock, frequency):
+        super().add_clock_constraint(clock, frequency)
+        clock.attrs["keep"] = "TRUE"
+
+    def create_missing_domain(self, name):
+        if name == "sync" and self.default_clk is not None:
+            m = Module()
+            if self.default_clk == "sys_clk0":
+                if not hasattr(self, "osc_div"):
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 and 512")
+                if not isinstance(self.osc_div, int) or self.osc_div < 2 or self.osc_div > 512:
+                    raise ValueError("OSC divider (osc_div) must be an integer between 2 and 512, not {!r}".format(self.osc_div))
+                if not hasattr(self, "osc_freq"):
+                    raise ValueError("OSC frequency (osc_freq) must be an integer between 2100000 and 80000000")
+                if not isinstance(self.osc_freq, int) or self.osc_freq < 2100000 or self.osc_freq > 80000000:
+                    raise ValueError("OSC frequency (osc_freq) must be an integer between 2100000 and 80000000, not {!r}".format(self.osc_freq))
+                clk_i = Signal()
```

Comment:
This doesn't look right--you aren't using `osc_div` and `osc_freq` anywhere in the instantiation and `qlal4s3b_cell_macro` is a QuickLogic primitive. Do GateMate FPGAs have any internal oscillator at all, like RC oscillator or ring oscillator? If not then the `create_missing_domain` function should only use an async reset synchronizer, or possibly just be absent.

Reply:
From what I have read on their documentation for the CCGM1A1 Chip, there are no oscillators(ring or RC) on the chip.

Reply:
Then you don't need any implementation of `create_missing_domain` or `default_clk_constraint` at all.

Reply:
Removed `create_missing_domain` and `default_clk_constraint`.